### PR TITLE
Add readOnlyRootFilesystem to security context (#2771)

### PR DESCRIPTION
**Pull request by david-yu**
* Bring in PR #2771 and run with CI which fixes https://github.com/hashicorp/consul-k8s/issues/751 

Changes proposed in this PR:
-
-

How I've tested this PR:

How I expect reviewers to test this PR:

Checklist:
- [ ] Tests added
- [ ] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Review by pglass — APPROVED** [review]

File: .changelog/2781.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
```
Comment (david-yu) [inline .changelog/2781.txt:1]: @mr-miles Does this sound right?

Comment (mr-miles) [reply .changelog/2781.txt]: 👍

**Comment by david-yu** [comment]
@mr-miles It looks like the PRs were able to pass more tests after adding the `tmp` volume mount to the server statefulset as well. Could you perhaps provide us some details on how this tmp volume is used, I'm not exactly sure why that fixes things, is it because the logs of raft logs are now written to tmp instead of folders we can write to?

**Comment by mr-miles** [comment]
With the readonlyrootvolume set to false, the server could write to
anywhere on the filesystem.

With it set to true, the container can only write to volumes that you’ve
mounted.

So my guess is that the server had always even writing some files into /tmp
(maybe ones you don’t care about, as all the important bits are persisted
in datadir) but now you need to make it explicit with the mount.

(If you can shell into the running pod you can see what files are there.
I’m afk otherwise I’d look)

**Review by curtbushko — COMMENTED** [review]
LGTM!

**Comment by david-yu** [comment]
@mr-miles Thanks again for your contribution, will go ahead and merge and backport to 1.2.x only to reduce blast radius.

**Review by curtbushko — APPROVED** [review]

**Comment by t-eckert** [comment]
Update, this is breaking Dataplane and all acceptance tests because Dataplane writes to disk when it starts up: https://github.com/hashicorp/consul-dataplane/blob/da6bf688df5db0e93c8b5c1e9e47227a34e51ea1/pkg/envoy/proxy.go#L337

This is read by Envoy for doing bootstrapping.

**Comment by mr-miles** [comment]
I couldn't work out how the consul-k8s helm chart is being utilised by the
integration tests, but FWIW I think the tests need to be updated to mount a
/tmp volume and so match the helm chart:

https://github.com/hashicorp/consul-dataplane/blob/da6bf688df5db0e93c8b5c1e9e47227a34e51ea1/integration-tests/helpers/dataplane.go#L76

and maybe also

https://github.com/hashicorp/consul-dataplane/blob/da6bf688df5db0e93c8b5c1e9e47227a34e51ea1/integration-tests/helpers/server.go#L82

Then dataplane can write to /tmp in line with what it is expecting

**Comment by david-yu** [comment]
@mr-miles It looks like this was likely not the culprit that broke our tests. I'll re-raise the PR once our pipelines have stabilized.

**Comment by mr-miles** [comment]
Thanks David - no problem (glad not to be implicated!), happy to help if it
is still causing grief after that

**Pull request merged**